### ISSUE TITLE
Crash when depaginating

### ISSUE THREAD

**Issue by ryan99fl** [rethinkdb/rethinkdb#5256]
Reproducible immediately after rethinkdb restart.
Turning off depagination or setting pageLimit to 0 prevents crash.

rethinkdb 2.2.2~0trusty (GCC 4.8.2)
Ubuntu 14.04.3 LTS -- Release 14.04 -- trusty

Core dump: [core.zip](https://github.com/rethinkdb/rethinkdb/files/79149/core.zip)

Data Explorer:
```javascript
const CW_API_URL = "https://my/restful/api";
const CW_COMPANY_ID = "mycompany";
const CW_KEY_PUBLIC = "foo";
const CW_KEY_PRIVATE = "bar";

r.http(CW_API_URL, {
  method: 'GET',
  auth: {
    type: 'basic',
    user: CW_COMPANY_ID + "+" + CW_KEY_PUBLIC,
    pass: CW_KEY_PRIVATE
  },
  params: {
    pageSize: 100
  },
  page: 'link-next',
  pageLimit: 3
});
```

Console Dump:
```
Server ready, "baz_uzt" d373dd09-fbba-4b90-a6d6-b4a1910bc276
Version: rethinkdb 2.2.2~0trusty (GCC 4.8.2)
error: Error in src/errors.cc at line 104:
error: Segmentation fault.
error: Backtrace:
addr2line: 'rethinkdb': No such file
addr2line: 'rethinkdb': No such file
error: Tue Jan  5 21:34:22 2016
 
       1 [0xa5d530]: backtrace_t::backtrace_t() at 0xa5d530 (rethinkdb)
       2 [0xa5d8c3]: format_backtrace(bool) at 0xa5d8c3 (rethinkdb)
       3 [0xc9f425]: report_fatal_error(char const*, int, char const*, ...) at 0xc9f425 (rethinkdb)
       4 [0xc9f5ac]: generic_crash_handler(int) at 0xc9f5ac (rethinkdb)
       5 [0x7fdec93e7340]: /lib/x86_64-linux-gnu/libpthread.so.0(+0x10340) [0x7fdec93e7340] at 0x7fdec93e7340 (/lib/x86_64-linux-gnu/libpthread.so.0)
       6 [0x9de010]: shared_buf_t::size() const at 0x9de010 (rethinkdb)
       7 [0x889e4f]: datum_string_t::size() const at 0x889e4f (rethinkdb)
       8 [0x9253a0]: ql::datum_serialized_size(ql::datum_t const&, ql::check_datum_serialization_errors_t, std::vector<ql::size_tree_node_t, std::alloca
tor<ql::size_tree_node_t> >*) at 0x9253a0 (rethinkdb)
       9 [0x9251a5]: ql::datum_object_serialized_size(ql::datum_t const&, ql::check_datum_serialization_errors_t, std::vector<ql::size_tree_node_t, std:
:allocator<ql::size_tree_node_t> >*) at 0x9251a5 (rethinkdb)
       10 [0x9256c4]: ql::datum_serialized_size(ql::datum_t const&, ql::check_datum_serialization_errors_t, std::vector<ql::size_tree_node_t, std::alloc
ator<ql::size_tree_node_t> >*) at 0x9256c4 (rethinkdb)
       11 [0x9251a5]: ql::datum_object_serialized_size(ql::datum_t const&, ql::check_datum_serialization_errors_t, std::vector<ql::size_tree_node_t, std
::allocator<ql::size_tree_node_t> >*) at 0x9251a5 (rethinkdb)
       12 [0x9256c4]: ql::datum_serialized_size(ql::datum_t const&, ql::check_datum_serialization_errors_t, std::vector<ql::size_tree_node_t, std::alloc
ator<ql::size_tree_node_t> >*) at 0x9256c4 (rethinkdb)
       13 [0x925aea]: ql::datum_serialize(write_message_t*, ql::datum_t const&, ql::check_datum_serialization_errors_t) at 0x925aea (rethinkdb)
       14 [0xc9386a]: void serialize<(cluster_version_t)7>(write_message_t*, http_result_t const&) at 0xc9386a (rethinkdb)
       15 [0xc9997f]: http_job_t::worker_fn(read_stream_t*, write_stream_t*) at 0xc9997f (rethinkdb)
       16 [0xc8fa22]: spawner_run_t::main_loop() at 0xc8fa22 (rethinkdb)
       17 [0xc8f50d]: extproc_spawner_t::fork_spawner() at 0xc8f50d (rethinkdb)
       18 [0xc8f596]: extproc_spawner_t::extproc_spawner_t() at 0xc8f596 (rethinkdb)
       19 [0xba7744]: main_rethinkdb_porcelain(int, char**) at 0xba7744 (rethinkdb)
       20 [0x6fb074]: main+0x1f4 at 0x6fb074 (rethinkdb)
       21 [0x7fdec9033ec5]: __libc_start_main+0xf5 at 0x7fdec9033ec5 (/lib/x86_64-linux-gnu/libc.so.6)
       22 [0x7068e9]: rethinkdb() [0x7068e9] at 0x7068e9 ()
error: Exiting.
```

Install method:
```
source /etc/lsb-release && echo "deb http://download.rethinkdb.com/apt $DISTRIB_CODENAME main" | sudo tee /etc/apt/sources.list.d/rethinkdb.list
wget -qO- https://download.rethinkdb.com/apt/pubkey.gpg | sudo apt-key add -
sudo apt-get update
sudo apt-get install rethinkdb
```

**Comment by encryptio** [comment]
Same backtrace as #4793 and #5252

**Comment by ryan99fl** [comment]
Reproducible immediately after rethinkdb restart.
Turning off depagination prevents crash.
Setting pageLimit to 0 prevents crash.

rethinkdb 2.2.2~0trusty (GCC 4.8.2)
Ubuntu 14.04.3 LTS -- Release 14.04 -- trusty

Install method:
```
source /etc/lsb-release && echo "deb http://download.rethinkdb.com/apt $DISTRIB_CODENAME main" | sudo tee /etc/apt/sources.list.d/rethinkdb.list
wget -qO- https://download.rethinkdb.com/apt/pubkey.gpg | sudo apt-key add -
sudo apt-get update
sudo apt-get install rethinkdb
```

**Comment by ryan99fl** [comment]
@encryptio @danielmewes May wish to pull #4793 out of backlog... also reappeared in #5252 as noted above.  Added coredump to OP.  Please advise if any further information requested.  Would be happy to build patched code (with assistance) to test in our environment.

**Comment by ryan99fl** [comment]
Trying to see if this is a regression bug - crash dumps to console in 2.1.5+2~0trusty but not in 2.0.0+1~0 trusty.  Trying to nail down specific version where bug reappears.

**Comment by ryan99fl** [comment]
Crash dump does not occur in v2.0.5~0trusty and previous but reappears in v2.1.0+1~0trusty.

**Comment by Tryneus** [comment]
So, the only difference I see in the `extproc/http_*` code between the `v2.0.5` and `v2.1.0-1` tags is changing `cJSON` to `rapidjson`, which, admittedly, I had been scrutinizing when trying to track this down.  The commits were `58184697`, `d7f43591`, and `7286b9fb`, in file `extproc/http_job.cc`.

I don't see anything immediately wrong with this code, but I could be missing something obvious.  The bug, at its root, looks like an invalid reference to a temporary or freed shared buffer, which was possibly generated when parsing the HTTP response as JSON.

Thanks for helping to track this down, @ryan99fl!

**Comment by ryan99fl** [comment]
@Tryneus So it may be malformed JSON?  To what RFC or other does this parser adhere?  I'd be happy to help validate or test.

**Comment by Tryneus** [comment]
@ryan99fl - I doubt it's malformed JSON, although it's possible that bad error handling exhibits this crash - it would be very useful if we could see the HTTP request/response that resulted in the crash, but I understand that might take some work to obtain.

**Comment by Tryneus** [comment]
Also, the rapidjson library can be found [here](https://github.com/miloyip/rapidjson), and it states:

> RapidJSON should be in fully compliance with RFC7159/ECMA-404

**Comment by ryan99fl** [comment]
I verified that the API is providing properly formatted data.  Will attempt to gather headers.

**Comment by ryan99fl** [comment]
The API in question only responds to https requests.  I do not have the option to obtain a private key from the remote API server.  Suggestions and I will provide any further info as requested.

**Comment by ryan99fl** [comment]
Discovered that the offending response from the API is as follows:
```
Status
400 Bad Request Show explanation Loading time: 54
Request headers 
User-Agent: Mozilla/5.0 (Windows NT 10.0; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/47.0.2526.106 Safari/537.36
Authorization: foo bar baz
Accept: */*
Accept-Encoding: gzip, deflate, sdch
Accept-Language: en-US,en;q=0.8
Response headers 
Cache-Control: private 
Content-Type: application/json; charset=utf-8 
Date: Thu, 07 Jan 2016 02:12:37 GMT 
Server: Microsoft-IIS/8.5 
X-AspNet-Version: 4.0.30319
X-Powered-By: ASP.NET
x-server-name: IP-0A0A071C
transfer-encoding: chunked
Connection: keep-alive
```
The offending JSON payload is:
```
{
  "code": "Security",
  "message": "SSL is required.",
  "errors": null
}
```
Version 2.0.5 returns a RqlRuntimeError in the Data Explorer output but does not crash or produce a core dump.  Version 2.1.0 crashes as noted above.

**Comment by Tryneus** [comment]
@ryan99fl - I can't reproduce it consistently, but I did observe the crash once, and got a core dump that I'm walking through to see if I can find where the missing buffer is coming from.  Thanks for all your help, and hopefully we'll have a fix soon.

**Comment by Tryneus** [comment]
The fix is up in review 3411.  It appears that this crash is likely to show up if the `r.http` workload sometimes has `Link` headers, and sometimes not.

**Comment by sagivf** [comment]
@Tryneus @ryan99fl This bug is very curtail for me - thanks for cracking down on it.
@Tryneus What do you mean in review 3411?
I'm not exactly clear on the problem, Is there anything I can do to avoid it? Any workaround?
How complex is the problem? Do you believe a solution can be made/released in a short time frame?

**Comment by danielmewes** [comment]
@sagivf I'm not sure if there's a workaround right now (@Tryneus will know better).
The review number is for our internal code review system.

We're going to include this fix in RethinkDB 2.2.3, which should be out by coming Monday.

**Comment by Tryneus** [comment]
@sagifv - unfortunately the best workaround I can think of requires that the HTTP responses never contain `Link` headers.  Sorry this is causing problems for you, but the fix should be available soon.

**Comment by Tryneus** [comment]
The fix has been approved and merged to `next` in commit 6d73f4a, and cherry-picked into `v2.2.x` in commit ddda52c.  Will be in release 2.2.3.

**Comment by sagivf** [comment]
@Tryneus @danielmewes I'm happy to say we upgraded to 2.2.x and all seems well! Thank you again for the quick turnaround and help, I'm using `r.http` pretty heavily and aside from this issue it seems to be doing a great Job. Keep up the good work :)

**Comment by ryan99fl** [comment]
@sagivf did you build locally from the patched 2.2.x branch? I can't believe that the fix has already made it into the pre compiled binaries, but please correct me if I am mistaken.

**Comment by sagivf** [comment]
@ryan99f l had help from @dalanmiller In creating a deb package with the fix.
